### PR TITLE
Remove dead code concernin TypeParameterType/Symbol from Microsoft.CSharp

### PR DESCRIPTION
**Pull request by JonHanna**
Most of this is code that relates to possibilities when compiling a generic method (including methods within generic types) that aren't relevant with the concrete types dynamic code works with.

* Remove cases concerning `TypeParameterType` conversions

Only relevant when compiling generic methods, not in dynamic code.

Entails removal of `ERR_TypeVarCantBeNull`. Contributes to #22470

* Remove case for `TypeParameterType` in `AllPossibleInterfaces`

Since both call sites ensure the type is a struct, class or interface type, it can only be called with an `AggregateType` or `NullableType`

* Remove `_bHasRefBound`, `_bHasValBound` and `_pDeducedBaseClass`

Only ever false, false and null respectively.

* Remove base and boxing type parameter conversions.

Only relevant when compiling generic methods, so never hit in dynamic code.

* Don't consider type parameter types in `GetRefEqualSigs`

Only comes up when statically compiling generic methods.

* Remove unnecessary `goto`s

* Don't consider type parameter types in `HasImplicitReferenceConversion`

Only comes up when static compiling generic methods.

* Reduce repeated checks within `HasImplicitReferenceConversion`

(Some of these will become better still when some methods that only return true for `AggregateType` are made sealed and virtual within it)

* Remove unnecessary identity check form `HasImplicitReferenceConversion`.

Never true when it is called, and if it was true the method should return true anyway.

* Move check that `pDest` is `AggregateType`

As all true-returning branches depend on it in some way, and it simplifies the rest.

* Replace if ladder on types of `AggregateType` into switch on symbol's `AggKind`

* Remove `TypeParameterType` case from `MemberLookup.Lookup`

Cannot happen outside of static compilation of generic method.

* Remove type parameter case from `SearchSingleType`

Cannot happen in dynamic code.

* Remove `TypeVarsAllowed` and `ExtensionCall` from `MemLookFlags`

Unused

* Remove `GetInterfaceBounds` from `TypeParameterType` and `TypeParameterSymbol`

Always null and no longer used.

* Remove `GetEffectiveBaseClass` from `TypeParameterType` & `TypeParameterSymbol`

Always null, and unused. (Removing either still results in null being the value that would be found, or removing a case that would throw if it was hittable).

### CODE REVIEW

**Comment by JonHanna** [comment]
Test OSX x64 Debug Build
Timeout, case of #23201.

**Comment by JonHanna** [comment]
Test OSX x64 Debug Build

**Comment by karelz** [comment]
@VSadov @OmarTawfik can you please review? it is 3 weeks without any response :(

**Comment by OmarTawfik** [comment]
@karelz @JonHanna Thanks for submitting/pinging. I'll get to this ASAP. going over a couple of bugs first :(

File: src/Microsoft.CSharp/src/Microsoft/CSharp/RuntimeBinder/Semantics/ExplicitConversion.cs [hunk 1]
```diff
@@ -141,12 +141,7 @@ public bool Bind()
                         return false; // Can't convert to a method group or anon method.
                     case TypeKind.TK_NullType:
                         return false;  // Can never convert TO the null type.
-                    case TypeKind.TK_TypeParameterType:
-                        if (bindExplicitConversionToTypeVar())
-                        {
-                            return true;
-                        }
-                        break;
+
```
Comment (OmarTawfik) [inline src/Microsoft.CSharp/src/Microsoft/CSharp/RuntimeBinder/Semantics/ExplicitConversion.cs]: nit: remove empty line?

File: src/microsoft.csharp/src/microsoft/csharp/runtimebinder/semantics/types/typeparametertype.cs [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Diagnostics;
+
+namespace Microsoft.CSharp.RuntimeBinder.Semantics
+{
+    /////////////////////////////////////////////////////////////////////////////////
+
+    internal sealed class TypeParameterType : CType
```
Comment (OmarTawfik) [inline src/microsoft.csharp/src/microsoft/csharp/runtimebinder/semantics/types/typeparametertype.cs]: If I understand this PR correctly, why keeping `TypeParameterType` and `TypeParameterSymbol` classes? shouldn't they removed with their uses?

Comment (JonHanna) [reply src/microsoft.csharp/src/microsoft/csharp/runtimebinder/semantics/types/typeparametertype.cs]: They're still used in defining types and methods. We can still have a generic method on a dynamic object called, or a dynamic object passed to a generic method, so it needs to start with a TypeParameterType in the definition and then works out the type that parameter needs to be in the call. It also creates generic types open-generic first (changing from that approach is perhaps one possible way to stop the problem that circular type definitions, but I'm not clear on if that's a good idea or not yet).

What we can't have, unless I'm missing something, is casts between type parameters as one can if compiling `(T)value` within a generic method, calling a generic with certain constraints from another generic, etc. which a lot of what this removes relates to. (With a little bit of tidying up of code near where removals had happened).

File: src/Microsoft.CSharp/src/Microsoft/CSharp/RuntimeBinder/Semantics/ExplicitConversion.cs (hunk 1 above)
Comment (JonHanna) [reply src/Microsoft.CSharp/src/Microsoft/CSharp/RuntimeBinder/Semantics/ExplicitConversion.cs]: Do we have a guideline on spacing between cases? I normally have a space before all but the first in a switch, and have been doing that most of the time I touch on cases as part of another change.

**Comment by karelz** [comment]
@OmarTawfik what is the next step? It's been another 3.5 weeks :(

File: src/Microsoft.CSharp/src/Microsoft/CSharp/RuntimeBinder/Semantics/ExplicitConversion.cs (hunk 1 above)
Comment (OmarTawfik) [reply src/Microsoft.CSharp/src/Microsoft/CSharp/RuntimeBinder/Semantics/ExplicitConversion.cs]: It is about consistency. Either we have a single empty line before all but the first in a switch, or none at all.

**Comment by OmarTawfik** [comment]
@karelz I already left one style comment. Otherwise, LGTM :)

**Comment by karelz** [comment]
@OmarTawfik ah, it was not clear - if you can mark it as "Approved", that would be great ;)
Do you want to block on the style comment, or is it ready for merge?

If it is ready for merge, please merge it ...

**Comment by OmarTawfik** [comment]
@dotnet-bot test this

**Comment by JonHanna** [comment]
@dotnet-bot test this, please.

**Comment by OmarTawfik** [comment]
[Probably unrelated failure](https://ci3.dot.net/job/dotnet_corefx/job/master/job/windows-TGroup_netfx+CGroup_Release+AGroup_x86+TestOuter_false_prtest/3820)
Trying another time to compare results
@dotnet-bot test NETFX x86 Release Build

**Comment by OmarTawfik** [comment]
Thanks @JonHanna!

**Pull request merged**